streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java

```diff
@@ -86,12 +87,29 @@ private void runOnce(final long nowMs) {
 
             if (currentTask == null) {
                 currentTask = taskManager.assignNextTask(DefaultTaskExecutor.this);
-            } else {
-                // if a task is no longer processable, ask task-manager to give it another
-                // task in the next iteration
-                if (currentTask.isProcessable(nowMs)) {
+            }
+
+            if (currentTask != null) {
```

Comment:
Just throwing ideas here.
Can we not move those tests into the `assignNextTask()` method, so that we guarantee that `assignNextTask()` never returns null. But probably, we also need to consider the case where no new task can be assigned.
Shouldn't we use a condition variable or similar in the case where a task cannot be assigned to avoid busy waiting?

Reply:
Let's discuss that and put it into a separate PR. WDYT?

Reply:
Yes, I think this is good for a separate PR. Also not really related to punctuation anymore.

Reply:
I checked this out and we need to be a bit careful here to not have the task executors to sleep when work is available. We had several bugs like this in state updater. We have to make sure to signal on the condtion variable any time a task becomes "processable". Here are some situations where a task becomes processable:

 - Task is unassigned from another TaskExecutor.
 - Task state is changed (should only happen inside when a task is locked inside the polling phase).
 - When tasks are unlocked.
 - When tasks are added.
 - New records available.
 - A task is resumed.

So in summary, we need to awake the task executors after every poll phase, inside `StreamThread.signalResume`, `DefaultTaskManager.unassignTask`, `DefaultTaskManager.unlockTasks`, `DefaultTaskManager.addTask`.